File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.js
/*
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteMiddle = function(head) {
    
    const dfs = (node, depth) => {
        let totalDepth = 0;
        if (node.next != null) {
            totalDepth = dfs(node.next, depth+1);   
        } else {
            totalDepth = depth;   
        }
        
        if(Math.ceil(totalDepth / 2) == depth + 1) {
            const temp = node.next;
            node.next = temp.next;
            temp.next = null;
        }
        
        return totalDepth;
    }
    const root = new ListNode(0, head);
    dfs(root, -1);
    return root.next;
};

File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        total = 1
        cur = head
        while cur.next:
            total += 1
            cur = cur.next

        half = total // 2
        # print(total, half)
        cur_index = 0
        cur = head
        while cur_index < half - 1:
            cur_index += 1
            cur = cur.next
        if cur.next and cur.next.next:
            cur.next = cur.next.next
        elif cur.next:
            cur.next = None
        else:
            head = None
        return head
